state/upgrades.go

```diff
@@ -1405,30 +1405,42 @@ func DeleteCloudImageMetadata(st *State) error {
 // CopyMongoSpaceToHASpaceConfig copies the Mongo space name from
 // ControllerInfo to the HA space name in ControllerConfig.
 // This only happens if the Mongo space state is valid, it is not empty,
-// and if the is no value already set for the HA space name.
-func CopyMongoSpaceToHASpaceConfig(st *State) error {
-	info, err := st.ControllerInfo()
-	if err != nil {
-		return errors.Annotate(err, "cannot get controller info")
-	}
-
-	ms := info.MongoSpaceName
-	if info.MongoSpaceState != MongoSpaceValid || ms == "" {
-		return nil
+// and if there is no value already set for the HA space name.
+// The old keys are then deleted from ControllerInfo.
+func MoveMongoSpaceToHASpaceConfig(st *State) error {
+	// Holds Mongo space fields removed from controllersDoc.
+	type controllersUpgradeDoc struct {
+		MongoSpaceName  string `bson:"mongo-space-name"`
+		MongoSpaceState string `bson:"mongo-space-state"`
 	}
+	var doc controllersUpgradeDoc
 
-	settings, err := readSettings(st.db(), controllersC, controllerSettingsGlobalKey)
+	controllerColl, controllerCloser := st.db().GetRawCollection(controllersC)
+	defer controllerCloser()
+	err := controllerColl.Find(bson.D{{"_id", modelGlobalKey}}).One(&doc)
 	if err != nil {
-		return errors.Annotate(err, "cannot get controller config")
+		return errors.Annotate(err, "retrieving controller info doc")
 	}
 
-	if _, ok := settings.Get(controller.JujuHASpace); ok {
-		return nil
+	if doc.MongoSpaceState == "valid" && doc.MongoSpaceName != "" {
+		settings, err := readSettings(st.db(), controllersC, controllerSettingsGlobalKey)
+		if err != nil {
+			return errors.Annotate(err, "cannot get controller config")
+		}
+
+		if _, ok := settings.Get(controller.JujuHASpace); !ok {
```

Comment:
Probably a comment here that
// if JujuHASpace is already set, don't override it, just let it be
maybe even a Debugf that its already set, so we won't overwrite it with the value from  MongoSpaceState?